Add relativeChildBounds, and use it for position inputs

volfied opened the pull request:
Will address #2613, and show layer position in relation to it's owner artboard.

 - Added `Bounds.prototype.relativeTo` method which returns a new bounds object with updated position in relation to the given bounds object.
 - Added `relativeChildBounds` cached property that calculates the bounds of a layer relative to it's owner artboard.
 - Added `selectedRelativeChildBounds` cached property that grabs all the selected layer's relativeChildBounds.
 - Add `relative` flag to `actions.transform.setPosition` to use relative child bounds during transform calculations, subsequently added it to `_getMoveLayerActions` as well.

volfied commented:
@chadrolfs @ktaki This is a high risk change, can you guys test out swap and transform features a bit in this branch when you have time?

pineapplespatula commented on src/js/actions/transform.js:
```diff
@@ -381,9 +385,9 @@ define(function (require, exports) {
             },
             layerOneActions = _getMoveLayerActions
-                .call(this, document, layers.get(0), newPositions.get(0), payload.positions),
+                .call(this, document, layers.get(0), newPositions.get(0), payload.positions, false),
```
why are you specifying false, instead of handling relative === undefined in the function

volfied replied:
Because I want that flag to be explicit across the codebase, just coding defensively.

pineapplespatula replied:
it would be more defensive to add a line to your function 
    if (relative === undefined) relative = false;

iwehrman replied:
Quasi-related: this is the sort of argument that might be better as a property of an `options` arguments.

chadrolfs commented:
on attempting to load this branch, I get 
`Uncaught TypeError: Cannot read property 'SHAPE' of undefined  rectangle.js:51`

volfied commented:
Chad's error was related to his bower link being out of date.

I added a new commit (6fb4e0c) which allows us to account for nesting changes when you setPosition of a layer into another artboard. This has a small cosmetic bug where if you drag a layer using mouse into another artboard, it's universal coordinates flash for a brief second before it being re-rendered. I'll be working on that, but I think it's good to go to master this way (barring all bugs you guys find)

iwehrman commented on src/js/actions/transform.js:
```diff
@@ -334,7 +337,8 @@ define(function (require, exports) {
                     target: documentLib.referenceBy.id(document.id)
-                }
+                },
+                synchronous: false
```
Can you add a comment here explaining why this flag is used?

iwehrman replied:
Still want this comment :)

iwehrman commented on src/js/actions/transform.js:
```diff
@@ -321,7 +324,7 @@ define(function (require, exports) {
      * @return {Promise}
      */
-    var setPosition = function (document, layerSpec, position) {
+    var setPosition = function (document, layerSpec, position, relative) {
```
Missing JSDoc on the new param.

iwehrman commented on src/js/actions/transform.js:
```diff
@@ -1044,10 +1050,18 @@ define(function (require, exports) {
         }, this);
 
+        _moveToArtboardHandler = function () {
+            var appStore = this.flux.store("application"),
```
Since you're in here touching this, it would probably be better to make this an action that transfers to `resetIndex`.

iwehrman commented:
Looks great, seems to work well. Back to @volfied. A deeper testing pass would still be helpful though, @chadrolfs or @ktaki.

volfied commented:
Rebased, addressed reviews, and made changes to `_getMoveLayerActions` so it returns the action just for the group layer. Problem was, if we moved a group with older code, we would move it's children to another artboard, but the group itself would not move since those layers got piecemeal translated.

Also, debounce handler has a 200 ms delay for "moveToArtboard" because we get one for each layer that's being transferred, and resetting index mid action breaks things.

volfied commented:
@iwehrman Back to you.

iwehrman commented on Gruntfile.js:
```diff
@@ -106,12 +106,11 @@ module.exports = function (grunt) {
-                "!**/*.ogg",
-                "!**/*.otf",
-                "!**/*.png"
+                "*.js",
```
I don't like the idea of not checking the svg files as well as the other text files in the root directory (the README, LICENSE, Gruntfile, etc.). How much time does this check actually save? More than 1s?

iwehrman commented:
Still missing a comment, plus complaints about the lint changes.

volfied commented:
@iwehrman added a comment, and also realized that (using unix `time` command), providing specific extensions to look for is much faster than giving it a whole directory. So lintspaces now checks the same amount of files but does it about 30 times faster.

iwehrman commented:
Zowie! I'm sold.
